cupy/lib/_function_base.py

```diff
@@ -0,0 +1,122 @@
+import warnings
+
+import cupy
+from cupy._core import internal
+
+
+def insert(arr, obj, values, axis=None):
+    """Insert values along the given axis before the given indices.
+
+    Parameters
+    ----------
+    arr : cupy.ndarray
+        Input array.
+    obj : int, slice or sequence of ints
+        Object that defines the index or indices before which `values` is
+        inserted.
+    values : cupy.ndarray
+        Values to insert into `arr`. If the type of `values` is different
+        from that of `arr`, `values` is converted to the type of `arr`.
+        `values` should be shaped so that ``arr[...,obj,...] = values``
+        is legal.
+    axis : int, optional
+        Axis along which to insert `values`.  If `axis` is None then `arr`
+        is flattened first.
+
+    Returns
+    -------
+    out : cupy.ndarray
+        A copy of `arr` with `values` inserted.  Note that `insert`
+        does not occur in-place: a new array is returned. If
+        `axis` is None, `out` is a flattened array.
+
+    Notes
+    -----
+    Note that for higher dimensional inserts `obj=0` behaves very different
+    from `obj=[0]` just like `arr[:,0,:] = values` is different from
+    `arr[:,[0],:] = values`.
+
+    .. seealso:: :func:`numpy.insert`
+    """
+
+    arr = cupy.asarray(arr)
+    ndim = arr.ndim
+    arrorder = 'F' if arr.flags.fnc else 'C'
+    if axis is None:
+        if ndim != 1:
+            arr = arr.ravel()
+        # needed for np.matrix, which is still not 1d after being ravelled
+        ndim = arr.ndim
+        axis = ndim - 1
+    else:
+        axis = internal._normalize_axis_index(axis, ndim)
+    slobj = [slice(None)]*ndim
+    N = arr.shape[axis]
+    newshape = list(arr.shape)
+
+    if isinstance(obj, slice):
+        indices = cupy.arange(*obj.indices(N), dtype=cupy.intp)
+    else:
+        # need to copy obj, because indices will be changed in-place
+        indices = cupy.array(obj)
```

Comment:
Here you need to be careful, creating a cupy array from a single integer can be expensive. It is faster to just pass the raw integer to the Cuda kernel since there are buffers for this.